### PR TITLE
[DEV-2266] save price path

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by Yannick1712** [comment]
Add migration

**Comment by Yannick1712** [comment]
TODO: Steps that obliterate each other should be removed

File: src/subdomains/supporting/pricing/domain/entities/price.ts
Comment (davidleomay) [inline src/subdomains/supporting/pricing/domain/entities/price.ts:1]: When creating a price, it normally has a single step. So I would just create the first step in the `create` method (no additional params needed).
When joining the prices with the `join` method, the steps should be appended (and inverting steps should be removed).

File: src/subdomains/supporting/pricing/domain/entities/price.ts [hunk 1]
```diff
@@ -42,3 +57,32 @@ export class Price {
...
+
+  @ApiProperty()
+  to: string;
+
+  @ApiProperty()
+  price: number;
+
+  @ApiProperty({ type: Date })
+  timestamp: Date;
+
```
Comment (davidleomay) [inline src/subdomains/supporting/pricing/domain/entities/price.ts:76]: Is `{ type: Date }` required?

File: src/subdomains/supporting/pricing/domain/entities/price.ts [hunk 2]
```diff
@@ -20,25 +23,67 @@ export class Price {
...
+    priceSource?: PriceSource,
+  ): Price {
     const price = new Price();
 
     price.source = source;
     price.target = target;
     price.price = _price;
     price.isValid = _isValid;
     price.timestamp = _timestamp;
+    price.steps = priceSource ? [PriceStep.create(priceSource, source, target, _price, _timestamp)] : [];
```
[the remark this answers is not in the document]
Comment (Yannick1712) [reply src/subdomains/supporting/pricing/domain/entities/price.ts:42]: Hm ok, makes it a little bit more complicate

File: src/subdomains/core/sell-crypto/process/services/buy-fiat-preparation.service.ts [hunk 3]
```diff
@@ -188,7 +188,7 @@ export class BuyFiatPreparationService {
         const price = await this.pricingService.getPrice(asset, currency, false);
 
         await this.buyFiatRepo.update(
-          ...entity.setOutput(price.convert(entity.inputReferenceAmountMinusFee), currency, price.steps),
+          ...entity.setOutput(price.convert(entity.inputReferenceAmountMinusFee), currency, price?.steps),
```
Comment (davidleomay) [inline src/subdomains/core/sell-crypto/process/services/buy-fiat-preparation.service.ts:191]: This `?` is useless 🤔

**Review by davidleomay — APPROVED** [review]

**Comment by TaprootFreak** [comment]
kann das auch aktiviert werden?

**Comment by Yannick1712** [comment]
Nee noch nicht, erst Anfang kommender Woche, muss noch testen

File: src/subdomains/supporting/pricing/domain/entities/price.ts [hunk 4]
```diff
@@ -58,7 +58,7 @@ export class Price {
 
     const priceSteps = prices.map((p) => p.steps).flat();
     const filteredPriceSteps = priceSteps.filter((p1) =>
-      priceSteps.some((p2) => p1.source === p2.source && p1.to === p2.from && p1.from === p2.to),
+      priceSteps.every((p2) => p1?.source !== p2?.source || p1?.to !== p2?.from || p1?.from !== p2?.to),
     );
```
Comment (davidleomay) [inline src/subdomains/supporting/pricing/domain/entities/price.ts:62]:
```ts
    const filteredPriceSteps = priceSteps.filter((p1) =>
      !priceSteps.some((p2) => p1.source === p2.source && p1.to === p2.from && p1.from === p2.to),
    );
```
would be a bit simpler to understand 😉

File: src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts [hunk 5]
```diff
@@ -365,6 +365,7 @@ export class BuyCrypto extends IEntity {
     minFeeAmountFiat: number,
     totalFeeAmountChf: number,
     feeConstraints: BuyCryptoFee,
+    priceSteps: PriceStep[],
   ): UpdateResult<BuyCrypto> {
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts:369]: I think the price steps should be set during output amount calculation `calculateOutputReferenceAmount`

File: src/subdomains/core/sell-crypto/process/buy-fiat.entity.ts [hunk 6]
```diff
@@ -236,6 +236,7 @@ export class BuyFiat extends IEntity {
     fee: InternalFeeDto & FeeDto,
     minFeeAmountFiat: number,
     totalFeeAmountChf: number,
+    priceSteps: PriceStep[],
   ): UpdateResult<BuyFiat> {
```
Comment (davidleomay) [inline src/subdomains/core/sell-crypto/process/buy-fiat.entity.ts:240]: Same here

File: src/subdomains/supporting/payment/dto/transaction-helper/tx-fee-details.dto.ts [hunk 7]
```diff
@@ -1,6 +1,8 @@
+import { PriceStep } from 'src/subdomains/supporting/pricing/domain/entities/price';
 import { FeeDto, InternalFeeDto } from '../fee.dto';
 
 export interface TxFeeDetails {
   minVolume: number;
   fee: InternalFeeDto & FeeDto;
+  priceSteps: PriceStep[];
 }
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/dto/transaction-helper/tx-fee-details.dto.ts:8]: This could then be removed again

Comment (davidleomay) [reply src/subdomains/supporting/payment/dto/transaction-helper/tx-fee-details.dto.ts:8]: - For BuyFiat, always inputReferenceAsset is always the same as inputAsset
- For BuyCrypto, they might be different but the price source fir conversion from input- to inputReferenceAsset is our bank

File: src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts [hunk 8]
```diff
@@ -204,9 +207,14 @@ export class BuyCrypto extends IEntity {
 
   calculateOutputReferenceAmount(price: Price): this {
     this.outputReferenceAmount = price.convert(this.inputReferenceAmountMinusFee, 8);
+    this.addPriceSteps(price?.steps);
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts:210]: `?` can be removed

File: src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts [hunk 9]
```diff
@@ -204,9 +207,14 @@ export class BuyCrypto extends IEntity {
...
     return this;
   }
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts:212]: If inputAsset is not the same as inputReferenceAsset, add a price step (first one) with source Bank

File: src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts [hunk 10]
```diff
@@ -204,9 +207,14 @@ export class BuyCrypto extends IEntity {
...
 
+  addPriceSteps(steps: PriceStep[]): void {
+    this.priceStepsObject = [...this.priceStepsObject, ...(steps ?? [])];
+  }
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts:216]: This method is useless?

File: src/subdomains/core/sell-crypto/process/buy-fiat.entity.ts [hunk 11]
```diff
@@ -336,6 +342,10 @@ export class BuyFiat extends IEntity {
     return [this.id, update];
   }
 
+  addPriceSteps(steps: PriceStep[]): void {
+    this.priceStepsObject = [...this.priceStepsObject, ...(steps ?? [])];
+  }
```
Comment (davidleomay) [inline src/subdomains/core/sell-crypto/process/buy-fiat.entity.ts:347]: This method is useless?

**Review by davidleomay — APPROVED** [review]

**Pull request merged**